# Exercise 16: Build, train and evaluate a neural network
 
## SIB - Intelligent Systems for Bioinformatics

Bárbara Freitas PG55693

Build, train and evaluate a NN following this instructions:

- The training dataset has 32 features
- The task is binary classification
- Use the SGD optimizer
- Use the BinaryCrossEntropy loss
- Use the accuracy metric
- The model should contain:
- Dense layer 1
- ReLU activation layer 1
- Dense layer 2
- ReLU activation layer 2
- Output Dense layer
- Sigmoid activation layer
- The dense layers should reduce the number of units to half (except the last one)
- Train the NN for 100 epochs, with batch size of 16 with a learning rate of 0.01.
- Test the model with k fold cross validation (hint: use the functions implemented in class 8)

### Imports

In [1]:

import numpy as np
from si.data.dataset import Dataset
from si.neural_networks.neural_network import NeuralNetwork

from si.model_selection.cross_validate import k_fold_cross_validation
from si.neural_networks.activation import ReLUActivation, SigmoidActivation
from si.neural_networks.layers import DenseLayer
from si.metrics.accuracy import accuracy

from si.neural_networks.losses import BinaryCrossEntropy

from si.neural_networks.optimizers import SGD

In [15]:
# Data Preparation
# Creating the 32 random features as requested
n_features = 32
dataset = Dataset.from_random(n_samples=5000, n_features=n_features, n_classes=2)
print(f"Dataset generated: {dataset.X.shape[0]} samples with {dataset.X.shape[1]} features.")

#  Model Configuration
# Initialize the network with the hyperparameters defined in the instructions.
net = NeuralNetwork(
    epochs=100,
    batch_size=16,
    learning_rate=0.01,
    optimizer=SGD,           
    loss=BinaryCrossEntropy, 
    metric=accuracy,
    verbose=True  
)

# Architecture (32 -> 16 -> 8 -> 1)
# Reduce units by half at each dense layer (32 -> 16 -> 8 -> 1)

# Layer 1: 32 Inputs -> 16 Neurons
net.add(DenseLayer(16, (32,)))
net.add(ReLUActivation())                        

# Layer 2: 16 Inputs -> 8 Neurons
net.add(DenseLayer(8))                     
net.add(ReLUActivation())                        

# Layer 3: 8 Inputs -> 1 Neuron
net.add(DenseLayer(1))        # Final dense layer for binary classification             
net.add(SigmoidActivation())     # Sigmoid activation for the probability output                  

# Run Cross Validation
print("Running k-fold cross-validation...")

results = k_fold_cross_validation(
    model=net,
    dataset=dataset,
    cv=5
)


Dataset generated: 5000 samples with 32 features.
Running k-fold cross-validation...
Epoch 1/100 - loss: 2786.0285 - accuracy: 0.4920
Epoch 2/100 - loss: 2776.7615 - accuracy: 0.5048
Epoch 3/100 - loss: 2774.5614 - accuracy: 0.5145
Epoch 4/100 - loss: 2777.3064 - accuracy: 0.5002
Epoch 5/100 - loss: 2774.6971 - accuracy: 0.5068
Epoch 6/100 - loss: 2772.4283 - accuracy: 0.5152
Epoch 7/100 - loss: 2772.0633 - accuracy: 0.5148
Epoch 8/100 - loss: 2772.8590 - accuracy: 0.5098
Epoch 9/100 - loss: 2772.5197 - accuracy: 0.5108
Epoch 10/100 - loss: 2771.1055 - accuracy: 0.5178
Epoch 11/100 - loss: 2772.1035 - accuracy: 0.5142
Epoch 12/100 - loss: 2769.7233 - accuracy: 0.5098
Epoch 13/100 - loss: 2769.8412 - accuracy: 0.5242
Epoch 14/100 - loss: 2767.2518 - accuracy: 0.5255
Epoch 15/100 - loss: 2765.4831 - accuracy: 0.5308
Epoch 16/100 - loss: 2763.0244 - accuracy: 0.5343
Epoch 17/100 - loss: 2761.2274 - accuracy: 0.5390
Epoch 18/100 - loss: 2760.1589 - accuracy: 0.5383
Epoch 19/100 - loss: 275

In [16]:
print(f"Accuracy: {np.mean(results):.2f} (+/- {np.std(results):.2f})")
print("Cross-validation scores:", results)
print("Average cross-validation score:", np.mean(results))

Accuracy: 0.57 (+/- 0.04)
Cross-validation scores: [0.499, 0.565, 0.567, 0.613, 0.61]
Average cross-validation score: 0.5708


### Performance Analysis

The model achieved an average accuracy of 0.57±0.04 across the 5-fold cross-validation. The individual scores ranged from a low of 0.499 to a high of 0.613.

Given that the dataset was generated using Dataset.from_random, the data consists of Gaussian noise with no underlying correlation between the 32 input features and the target classes. Consequently, the average accuracy is close to 50%, which represents random guessing in a binary classification task. The slight deviation above 50% (0.57) is attributable to spurious correlations found by the network in the random noise within specific folds, or slight class imbalances in the random splits.

The standard deviation of 0.04 and the range of scores (from ~50% to ~61%) indicate that the model's performance is unstable. This confirms that the model is not learning robust features but is instead overfitting to the specific noise patterns present in the training set of each fold.

In conclusion, the Neural Network architecture functions correctly, but as expected, it cannot extract meaningful patterns from random noise. To observe proper learning convergence (high accuracy and low loss), this architecture should be applied to a dataset with real relationships beetween feautures and targets.